File: javascript/alphametics/alphametics.js
export const solve = (puzzle) => {
  // Parse the puzzle into operands and result
  const parts = puzzle.split('==');
  if (parts.length !== 2) return null;
  
  const rightSide = parts[1].trim();
  const leftSide = parts[0].trim();
  const operands = leftSide.split('+').map(op => op.trim());
  
  // Extract all unique letters from the puzzle
  const letters = new Set();
  for (const char of puzzle) {
    if (/[A-Z]/i.test(char)) {
      letters.add(char);
    }
  }
  
  // If we have more than 10 unique letters, it's impossible to solve
  if (letters.size > 10) return null;
  
  // Find the first letter of each operand and the result
  const firstLetters = new Set();
  for (const operand of operands) {
    if (operand.length > 0) {
      firstLetters.add(operand[0]);
    }
  }
  if (rightSide.length > 0) {
    firstLetters.add(rightSide[0]);
  }
  
  // Convert sets to arrays for easier manipulation
  const lettersArray = Array.from(letters);
  const firstLettersArray = Array.from(firstLetters);
  
  // Generate permutations and check for valid solutions
  return findSolution(lettersArray, firstLettersArray, operands, rightSide);
};

/**
 * Recursively try different digit assignments to find a valid solution
 */
function findSolution(letters, firstLetters, operands, result, assignment = {}, index = 0, usedDigits = new Set()) {
  // Base case: all letters have been assigned a digit
  if (index === letters.length) {
    // Check if the solution is valid
    return isValidSolution(operands, result, assignment) ? { ...assignment } : null;
  }
  
  const letter = letters[index];
  
  // Try each possible digit for the current letter
  for (let digit = 0; digit <= 9; digit++) {
    // Skip if this digit is already used
    if (usedDigits.has(digit)) continue;
    
    // Skip if this is a first letter and digit is 0
    if (firstLetters.includes(letter) && digit === 0) continue;
    
    // Assign the digit to the letter
    assignment[letter] = digit;
    usedDigits.add(digit);
    
    // Recursively try to assign digits to the remaining letters
    const solution = findSolution(letters, firstLetters, operands, result, assignment, index + 1, usedDigits);
    if (solution) return solution;
    
    // Backtrack
    delete assignment[letter];
    usedDigits.delete(digit);
  }
  
  return null;
}

/**
 * Check if the current assignment produces a valid equation
 */
function isValidSolution(operands, result, assignment) {
  // Convert operands and result to numbers based on the assignment
  const operandValues = operands.map(operand => {
    return convertToNumber(operand, assignment);
  });
  
  const resultValue = convertToNumber(result, assignment);
  
  // Check if the sum of operands equals the result
  return operandValues.reduce((sum, value) => sum + value, 0) === resultValue;
}

/**
 * Convert a string of letters to a number using the assignment
 */
function convertToNumber(str, assignment) {
  let result = 0;
  for (const char of str) {
    if (/[A-Z]/i.test(char)) {
      result = result * 10 + assignment[char];
    }
  }
  return result;
}

File: go/alphametics/alphametics.go
package alphametics

import (
	"errors"
	"fmt"
	"strings"
)

// Solve takes an alphametics puzzle and returns a map of letters to digits
// that satisfies the equation, or an error if no solution exists.
func Solve(puzzle string) (map[string]int, error) {
	// Parse the puzzle into terms and result
	terms, result, err := parsePuzzle(puzzle)
	if err != nil {
		return nil, err
	}

	// Get all unique letters in the puzzle
	letters := getUniqueLetters(terms, result)

	// Get the first letter of each term and the result
	firstLetters := getFirstLetters(terms, result)

	// Try all possible digit assignments
	solution, found := findSolution(letters, firstLetters, terms, result)
	if !found {
		return nil, errors.New("no solution found")
	}

	return solution, nil
}

// parsePuzzle parses the puzzle string into terms and result
func parsePuzzle(puzzle string) ([]string, string, error) {
	// Remove spaces and split by ==
	parts := strings.Split(strings.ReplaceAll(puzzle, " ", ""), "==")
	if len(parts) != 2 {
		return nil, "", fmt.Errorf("invalid puzzle format: %s", puzzle)
	}

	// Split the left side by +
	terms := strings.Split(parts[0], "+")
	result := parts[1]

	return terms, result, nil
}

// getUniqueLetters returns all unique letters in the puzzle
func getUniqueLetters(terms []string, result string) []rune {
	letterMap := make(map[rune]bool)

	// Add letters from terms
	for _, term := range terms {
		for _, char := range term {
			letterMap[char] = true
		}
	}

	// Add letters from result
	for _, char := range result {
		letterMap[char] = true
	}

	// Convert map to slice
	letters := make([]rune, 0, len(letterMap))
	for letter := range letterMap {
		letters = append(letters, letter)
	}

	return letters
}

// getFirstLetters returns the first letter of each term and the result
func getFirstLetters(terms []string, result string) map[rune]bool {
	firstLetters := make(map[rune]bool)

	// Add first letter of each term
	for _, term := range terms {
		if len(term) > 0 {
			firstLetters[rune(term[0])] = true
		}
	}

	// Add first letter of result
	if len(result) > 0 {
		firstLetters[rune(result[0])] = true
	}

	return firstLetters
}

// evaluateExpression calculates the value of a term given a letter-to-digit mapping
func evaluateExpression(expr string, mapping map[rune]int) (int, bool) {
	if len(expr) == 0 {
		return 0, true
	}

	// Check if the first digit is 0 (invalid for multi-digit numbers)
	if mapping[rune(expr[0])] == 0 && len(expr) > 1 {
		return 0, false
	}

	// Calculate the value
	value := 0
	for _, char := range expr {
		value = value*10 + mapping[char]
	}

	return value, true
}

// findSolution tries all possible digit assignments to find a solution
func findSolution(letters []rune, firstLetters map[rune]bool, terms []string, result string) (map[string]int, bool) {
	// If there are more than 10 letters, no solution is possible
	if len(letters) > 10 {
		return nil, false
	}

	// Create a map to store the current assignment
	assignment := make(map[rune]int)
	used := make([]bool, 10) // Track which digits are already used

	// Try all possible assignments recursively
	if backtrack(letters, 0, firstLetters, assignment, used, terms, result) {
		// Convert rune map to string map for the result
		solution := make(map[string]int)
		for letter, digit := range assignment {
			solution[string(letter)] = digit
		}
		return solution, true
	}

	return nil, false
}

// backtrack recursively tries all possible digit assignments
func backtrack(letters []rune, index int, firstLetters map[rune]bool, assignment map[rune]int, used []bool, terms []string, result string) bool {
	// If all letters have been assigned, check if the equation is satisfied
	if index == len(letters) {
		return checkEquation(terms, result, assignment)
	}

	letter := letters[index]
	isFirstLetter := firstLetters[letter]

	// Try each possible digit for this letter
	start := 0
	if isFirstLetter {
		start = 1 // First letters can't be 0
	}

	for digit := start; digit <= 9; digit++ {
		// Skip if this digit is already used
		if used[digit] {
			continue
		}

		// Assign this digit to the letter
		assignment[letter] = digit
		used[digit] = true

		// Recursively try to assign the next letter
		if backtrack(letters, index+1, firstLetters, assignment, used, terms, result) {
			return true
		}

		// Backtrack
		delete(assignment, letter)
		used[digit] = false
	}

	return false
}

// checkEquation checks if the current assignment satisfies the equation
func checkEquation(terms []string, result string, assignment map[rune]int) bool {
	sum := 0

	// Calculate the sum of all terms
	for _, term := range terms {
		value, valid := evaluateExpression(term, assignment)
		if !valid {
			return false
		}
		sum += value
	}

	// Calculate the value of the result
	resultValue, valid := evaluateExpression(result, assignment)
	if !valid {
		return false
	}

	// Check if the equation is satisfied
	return sum == resultValue
}
